mm2src/mm2_net/src/native_tls/builder.rs

```diff
@@ -0,0 +1,92 @@
+use crate::native_tls::TlsAcceptor;
+use hyper::server::conn::AddrIncoming;
+use rustls::ServerConfig;
+use std::sync::Arc;
+
+/// Builder for [`TlsAcceptor`]
+pub struct AcceptorBuilder<State>(State);
+
+/// State of a builder that needs a TLS client config next
+pub struct WantsTlsConfig(());
+
+impl AcceptorBuilder<WantsTlsConfig> {
+    #[inline]
+    /// Creates a new [`AcceptorBuilder`]
+    pub fn new() -> Self { Self(WantsTlsConfig(())) }
+
+    #[inline]
+    /// Passes a rustls [`ServerConfig`] to configure the TLS connection
+    pub fn with_tls_config(self, config: ServerConfig) -> AcceptorBuilder<WantsAlpn> {
+        AcceptorBuilder(WantsAlpn(config))
+    }
+
+    /// Use rustls [defaults][with_safe_defaults] without [client authentication][with_no_client_auth]
+    ///
+    /// [with_safe_defaults]: rustls::ConfigBuilder::with_safe_defaults
+    /// [with_no_client_auth]: rustls::ConfigBuilder::with_no_client_auth
+    pub fn with_single_cert(
+        self,
+        cert_chain: Vec<rustls::Certificate>,
+        key_der: rustls::PrivateKey,
+    ) -> Result<AcceptorBuilder<WantsAlpn>, rustls::Error> {
+        Ok(AcceptorBuilder(WantsAlpn(
+            ServerConfig::builder()
+                .with_safe_defaults()
+                .with_no_client_auth()
+                .with_single_cert(cert_chain, key_der)?,
+        )))
+    }
+}
+
+impl Default for AcceptorBuilder<WantsTlsConfig> {
+    fn default() -> Self { Self::new() }
+}
+
+/// State of a builder that needs a incoming address next
+pub struct WantsAlpn(ServerConfig);
+
+impl AcceptorBuilder<WantsAlpn> {
+    /// Configure ALPN accept protocols in order
+    pub fn with_alpn_protocols(mut self, alpn_protocols: Vec<Vec<u8>>) -> AcceptorBuilder<WantsIncoming> {
+        self.0 .0.alpn_protocols = alpn_protocols;
+        AcceptorBuilder(WantsIncoming(self.0 .0))
+    }
+
+    /// Configure ALPN to accept HTTP/2
+    pub fn with_http2_alpn(mut self) -> AcceptorBuilder<WantsIncoming> {
+        self.0 .0.alpn_protocols = vec![b"h2".to_vec()];
+        AcceptorBuilder(WantsIncoming(self.0 .0))
+    }
+
+    /// Configure ALPN to accept HTTP/1.0
+    pub fn with_http10_alpn(mut self) -> AcceptorBuilder<WantsIncoming> {
+        self.0 .0.alpn_protocols = vec![b"http/1.0".to_vec()];
+        AcceptorBuilder(WantsIncoming(self.0 .0))
+    }
+
+    /// Configure ALPN to accept HTTP/1.1
+    pub fn with_http11_alpn(mut self) -> AcceptorBuilder<WantsIncoming> {
+        self.0 .0.alpn_protocols = vec![b"http/1.1".to_vec()];
+        AcceptorBuilder(WantsIncoming(self.0 .0))
+    }
+
+    /// Configure ALPN to accept HTTP/2, HTTP/1.1, HTTP/1.0 in that order.
+    pub fn with_all_versions_alpn(mut self) -> AcceptorBuilder<WantsIncoming> {
+        self.0 .0.alpn_protocols = vec![b"h2".to_vec(), b"http/1.1".to_vec(), b"http/1.0".to_vec()];
```

Comment:
I think you can make these protocol values constant

Reply:
This is a fork. We should not make a diff in it unless it's very necessary.

Reply:
ok make sense. thanks